### PR TITLE
Refactor

### PR DESCRIPTION
**Pull request by Domochip**
Hi,
I finished my new version.
It's a very very big update.

### CODE REVIEW

**Comment by Domochip** [comment]
Hi,
Sorry, I'm not yet very comfortable with Merge and PR and made bad merge and revert it...
Wasted

**Comment by wardbox** [comment]
No worries, don't rush it!

**Comment by Domochip** [comment]
My final thought is about a Backup/Restore feature.
Do you think I should include it in this project or make a different one?

**Comment by wardbox** [comment]
What would the backup restore be backing up and restoring?

**Comment by Domochip** [comment]
Something like Backup-Library for 'FollowedArtists', 'SavedAlbums', 'SavedShows', 'SavedTracks'
And Backup-Playlist for playlists

**Comment by wardbox** [comment]
I like it, that'd be pretty sweet.

**Comment by Domochip** [comment]
ok, done, I implemented Backup/Restore of Playlists, FollowedArtists, SavedAlbums, SavedShows, SavedTracks

File: Spotishell/Public/Backup-Library.ps1 [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+        Path of the backup file you want to create
+    .PARAMETER Type
+        One or more items type to backup (All, FollowedArtists, SavedAlbums, SavedShows, SavedTracks)
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Backup-Library {
+    param (
+        [Parameter(Mandatory)]
+        [string]
```
Comment (wardbox) [inline Spotishell/Public/Backup-Library.ps1:20]:
```suggestion
        [ValidateScript({
            if( -Not ($_ | Test-Path) ){
                throw "File or folder does not exist"
            }
            return $true
        })]
```

File: Spotishell/Public/Backup-Library.ps1 [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+        $Path,
```
Comment (wardbox) [inline Spotishell/Public/Backup-Library.ps1:21]:
```suggestion
        [System.IO.FileInfo]$Path,
```

File: Spotishell/Public/Albums/Get-Album.ps1 [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Get-Album {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Albums/Get-Album.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Follow/Add-FollowedUser.ps1 [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Add-FollowedUser {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Follow/Add-FollowedUser.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Follow/Test-FollowedArtist.ps1 [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Test-FollowedArtist {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Follow/Test-FollowedArtist.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Follow/Remove-FollowedUser.ps1 [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+    #>
+function Remove-FollowedUser {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Follow/Remove-FollowedUser.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Follow/Test-FollowedPlaylist.ps1 [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+function Test-FollowedPlaylist {
+    param (
+        [Parameter(Mandatory)]
+        [ValidateNotNullOrEmpty()]
+        [string]
+        $PlaylistId,
+
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [array]
+        $UserIds,
```
Comment (wardbox) [inline Spotishell/Public/Follow/Test-FollowedPlaylist.ps1:26]:
```suggestion
        $UserId,
```

File: Spotishell/Public/Follow/Test-FollowedUser.ps1 [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Test-FollowedUser {
+  param (
+      [Parameter(Mandatory, ValueFromPipeline)]
+      [ValidateNotNullOrEmpty()]
+      [Alias('Id')]
+      [array]
+      $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Follow/Test-FollowedUser.ps1:24]:
```suggestion
      $Id,
```

File: Spotishell/Public/Library/Add-CurrentUserSavedAlbum.ps1 [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Add-CurrentUserSavedAlbum {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Add-CurrentUserSavedAlbum.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Add-CurrentUserSavedShow.ps1 [hunk 10]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Add-CurrentUserSavedShow {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Add-CurrentUserSavedShow.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Add-CurrentUserSavedTrack.ps1 [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Add-CurrentUserSavedTrack {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Add-CurrentUserSavedTrack.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Remove-CurrentUserSavedAlbum.ps1 [hunk 12]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Remove-CurrentUserSavedAlbum {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Remove-CurrentUserSavedAlbum.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Remove-CurrentUserSavedShow.ps1 [hunk 13]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Remove-CurrentUserSavedShow {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Remove-CurrentUserSavedShow.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Remove-CurrentUserSavedTrack.ps1 [hunk 14]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Remove-CurrentUserSavedTrack {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Remove-CurrentUserSavedTrack.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Test-CurrentUserSavedAlbum.ps1 [hunk 15]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Test-CurrentUserSavedAlbum {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Test-CurrentUserSavedAlbum.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Test-CurrentUserSavedShow.ps1 [hunk 16]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Test-CurrentUserSavedShow {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Test-CurrentUserSavedShow.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Library/Test-CurrentUserSavedTracks.ps1 [hunk 17]
```diff
@@ -0,0 +1,37 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Test-UserSavedTracks {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Library/Test-CurrentUserSavedTracks.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Playlists/Add-PlaylistItem.ps1 [hunk 18]
```diff
@@ -0,0 +1,54 @@
...
+    param (
+        [Parameter(Mandatory)]
+        [string]
+        $Id,
+    
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('ItemId')]
+        [array]
+        $ItemIds,
```
Comment (wardbox) [inline Spotishell/Public/Playlists/Add-PlaylistItem.ps1:33]:
```suggestion
        $ItemId,
```

File: Spotishell/Public/Playlists/Get-PlaylistItems.ps1 [hunk 19]
```diff
@@ -0,0 +1,55 @@
...
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Get-PlaylistItems {
+    param(
+        [Parameter(Mandatory)]
+        [string]
+        $Id,
+
+        [string]
+        $Fields,
```
Comment (wardbox) [inline Spotishell/Public/Playlists/Get-PlaylistItems.ps1:30]:
```suggestion
        $Field,
```

File: Spotishell/Public/Playlists/Remove-PlaylistItems.ps1 [hunk 20]
```diff
@@ -0,0 +1,52 @@
...
+#>
+function Remove-PlaylistItems {
+    param (
+        [Parameter(Mandatory)]
+        [string]
+        $Id,
+
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [array]
+        $Tracks,
```
Comment (wardbox) [inline Spotishell/Public/Playlists/Remove-PlaylistItems.ps1:32]:
```suggestion
        $Track,
```

File: Spotishell/Public/Restore-Library.ps1 [hunk 21]
```diff
@@ -0,0 +1,91 @@
...
+        Path of the backup file you want to restore
+    .PARAMETER Type
+        One or more items type to restore (All, FollowedArtists, SavedAlbums, SavedShows, SavedTracks)
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Restore-Library {
+    param (
+        [Parameter(Mandatory)]
+        [string]
```
Comment (wardbox) [inline Spotishell/Public/Restore-Library.ps1:20]:
```suggestion
        [ValidateScript({
            if( -Not ($_ | Test-Path) ){
                throw "File or folder does not exist"
            }
            return $true
        })]
```

File: Spotishell/Public/Restore-Library.ps1 [hunk 22]
```diff
@@ -0,0 +1,91 @@
...
+        $Path,
```
Comment (wardbox) [inline Spotishell/Public/Restore-Library.ps1:21]:
```suggestion
        [System.IO.FileInfo]$Path,
```

File: Spotishell/Public/Shows/Get-Show.ps1 [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Get-Show {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Shows/Get-Show.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Tracks/Get-Track.ps1 [hunk 24]
```diff
@@ -0,0 +1,38 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Get-Track {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Tracks/Get-Track.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Tracks/Get-TrackAudioFeature.ps1 [hunk 25]
```diff
@@ -0,0 +1,38 @@
...
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Get-TrackAudioFeature {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
+        [array]
+        $Ids,
```
Comment (wardbox) [inline Spotishell/Public/Tracks/Get-TrackAudioFeature.ps1:24]:
```suggestion
        $Id,
```

File: Spotishell/Public/Tracks/Get-TrackAudioFeature.ps1 [hunk 26]
```diff
@@ -0,0 +1,38 @@
...
+    .PARAMETER Ids
+        One or more Track Ids
+    .PARAMETER ApplicationName
+        Specifies the Spotify Application Name (otherwise default is used)
+#>
+function Get-TrackAudioFeature {
+    param (
+        [Parameter(Mandatory, ValueFromPipeline)]
+        [ValidateNotNullOrEmpty()]
+        [Alias('Id')]
```
Comment (wardbox) [inline Spotishell/Public/Tracks/Get-TrackAudioFeature.ps1:22]:
```suggestion
```

**Review by wardbox — COMMENTED** [review]
This is awesome, man. The only changes I suggested were to make plural parameters singular, it just takes the question of "is this one '-Id' or '-Ids'?" out of the user experience. The alias would work but we can just avoid aliasing altogether for it. Also some paths I suggest validating on entry. Otherwise, it's good to go, thank you so much for the contributions!

File: Spotishell/Public/Backup-Library.ps1 (hunk 1 above)
Comment (Domochip) [reply Spotishell/Public/Backup-Library.ps1:20]: this cmdlet create the backup, so there is no problem if the file doesn't exists.

File: Spotishell/Public/Restore-Library.ps1 (hunk 21 above)
Comment (Domochip) [reply Spotishell/Public/Restore-Library.ps1:20]: The first line of this function is Get-Content which will throw the right Exception if there is a problem with $Path (like missing file)

**Comment by Domochip** [comment]
Should be good now. I really think that current code is able to catch file path problems (Backup/Restore).

**Review by wardbox — APPROVED** [review]
Looks good

**Comment by Domochip** [comment]
thanks but there is still 3 conflicting files that prevent merge :-)

**Pull request merged**